src/katgpucbf/utils.py

```diff
@@ -184,6 +186,84 @@ def cancel(self) -> None:
+    # These thresholds are thumb-sucks.
+    if 0.0 <= seconds < 1e-3:  # Arbitrary threshold
+        return aiokatcp.Sensor.Status.NOMINAL
+    elif seconds < 5e-3:  # CBF-REQ-0203 specifies 5ms
+        return aiokatcp.Sensor.Status.WARN
+    else:
+        return aiokatcp.Sensor.Status.ERROR
+
+
+def _time_maxerror_status(seconds: float) -> aiokatcp.Sensor.Status:
+    # maxerror is an over-estimate (it makes very conservative assumptions
+    # about network asymmetry and skew in the clock source). Use conservative
+    # thresholds to avoid warnings when there isn't a stratum 1 time source
+    # in the same data centre. Experimentally, maximum error is < 10ms
+    # when synchronising between Cape Town and the Karoo.
+    if 0.0 <= seconds < 10e-3:
+        return aiokatcp.Sensor.Status.NOMINAL
+    elif seconds < 0.1:
+        return aiokatcp.Sensor.Status.WARN
+    else:
+        return aiokatcp.Sensor.Status.ERROR
+
+
+def _time_state_status(value: aiokatcp.ClockState) -> aiokatcp.Sensor.Status:
+    if value == aiokatcp.ClockState.OK:
+        return aiokatcp.Sensor.Status.NOMINAL
+    elif value == aiokatcp.ClockState.ERROR:
+        return aiokatcp.Sensor.Status.ERROR
+    else:
+        # Some form of leap second adjustment
+        return aiokatcp.Sensor.Status.WARN
+
+
+def add_time_sync_sensors(sensors: aiokatcp.SensorSet) -> asyncio.Task:
+    """Add a number of sensors to a device server to track time synchronisation.
+
+    This must be called with an event loop running. It returns a task that
+    keeps the sensors periodically updated.
+    """
+    mapping: dict[str, aiokatcp.Sensor] = {
+        "esterror": aiokatcp.Sensor(
+            float, "time.esterror", "Estimated time synchronisation error", units="s", status_func=_time_esterror_status
+        ),
+        "maxerror": aiokatcp.Sensor(
+            float,
+            "time.maxerror",
+            "Upper bound on time synchronisation error",
+            units="s",
+            status_func=_time_maxerror_status,
+        ),
+        "state": aiokatcp.Sensor(
+            aiokatcp.ClockState, "time.state", "Kernel clock state", status_func=_time_state_status
+        ),
+    }
+    for sensor in mapping.values():
+        sensors.add(sensor)
+
+    synchronised_sensor = aiokatcp.Sensor(
+        bool, "time.synchronised", "Whether the host clock is synchronised within tolerances"
+    )
+    sensors.add(synchronised_sensor)
+    updater = aiokatcp.TimeSyncUpdater(mapping)
+
+    async def run() -> None:
+        while True:
+            updater.update()
+            good = all(sensor.status != aiokatcp.Sensor.Status.ERROR for sensor in mapping.values())
+            synchronised_sensor.set_value(
+                good,
+                status=aiokatcp.Sensor.Status.NOMINAL if good else aiokatcp.Sensor.Status.ERROR,
+                timestamp=mapping["state"].timestamp,
+            )
+            await asyncio.sleep(1)
```

Comment:
Why 1 second specifically?

Reply:
Ah, thank you for catching this. This is why we have two reviewers!

Reply:
I don't recall that conversion. Why 5 seconds?

Reply:
The update process is fairly cheap (just a system call, not a bunch of networking to talk to a daemon), so I figure we might as well update it reasonably fast, and if people want a lower update rate they can use one of the rate-limiting sensor strategies.

Reply:
Maybe I'm getting confused with something else.
But is there a reason for selecting 1 second? Maybe just add a comment with the logic. The whole Chesterton's fence thing, so that a future reader of the code can see why it's there.